library/nist_kw.c

```diff
@@ -0,0 +1,710 @@
+        memset( output + KW_SEMIBLOCK_LENGTH + in_len, 0, padlen );
+    }
+    semiblocks = ( ( in_len + padlen ) / KW_SEMIBLOCK_LENGTH ) + 1;
+
+    s = 6 * ( semiblocks - 1 );
+
+    if( mode == MBEDTLS_KW_MODE_KWP
+        && in_len <= KW_SEMIBLOCK_LENGTH )
+    {
+        memcpy( inbuff, output, 16 );
+        ret = mbedtls_cipher_update( &ctx->cipher_ctx,
+                                     inbuff, 16, output, &olen );
+        if( ret != 0 )
+            goto cleanup;
+
+    }
+    else
+    {
+        /*
+         * Do the wrapping function W, as defined in RFC 3394 section 2.2.1
+         */
+        if( semiblocks < MIN_SEMIBLOCKS_COUNT )
+        {
+            ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+            goto cleanup;
+
+        }
+
+        /* Calculate intermediate values */
+        for( t = 1; t <= s; t++ )
+        {
+            memcpy( inbuff, A, KW_SEMIBLOCK_LENGTH );
+            memcpy( inbuff + KW_SEMIBLOCK_LENGTH, R2, KW_SEMIBLOCK_LENGTH );
+
+            ret = mbedtls_cipher_update( &ctx->cipher_ctx,
+                                         inbuff, 16, outbuff, &olen );
+            if( ret != 0 )
+                goto cleanup;
+
+            memcpy( A, outbuff, KW_SEMIBLOCK_LENGTH );
+            calc_a_xor_t( A, t );
+
+            memcpy( R2, outbuff + KW_SEMIBLOCK_LENGTH, KW_SEMIBLOCK_LENGTH );
+            R2 += KW_SEMIBLOCK_LENGTH;
+            if( R2 >= output + ( semiblocks * KW_SEMIBLOCK_LENGTH ) )
+                R2 = output + KW_SEMIBLOCK_LENGTH;
+        }
+    }
+    /*
+     * Set the output length as the number.
+     * If mode is MBEDTLS_KW_MODE_KWP and in_len < 8,
+     * Then after padding, the number of semiblocks in the output would be 2,
+     * Which means the output is 16 bytes, as expected.
+     */
+    *out_len = semiblocks * KW_SEMIBLOCK_LENGTH;
+
+cleanup:
+
+    if( ret != 0)
+    {
+        memset( output, 0, semiblocks * KW_SEMIBLOCK_LENGTH );
+    }
+    mbedtls_platform_zeroize( inbuff, KW_SEMIBLOCK_LENGTH * 2 );
+    mbedtls_platform_zeroize( outbuff, KW_SEMIBLOCK_LENGTH * 2 );
+    mbedtls_cipher_finish( &ctx->cipher_ctx, NULL, &olen );
+    return( ret );
+}
+
+/*
+ * W-1 function as defined in RFC 3394 section 2.2.2:
+ */
+static int unwrap( mbedtls_nist_kw_context *ctx,
+                   const unsigned char *input, size_t semiblocks,
+                   unsigned char *output, size_t* out_len )
+{
+    int ret = 0;
+    const size_t s = 6 * ( semiblocks -1 );
+    size_t olen;
+    uint64_t t = 0;
+    unsigned char outbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char inbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char *R = output + ( semiblocks -1 ) * KW_SEMIBLOCK_LENGTH;
+    unsigned char *A = output;
+    *out_len = 0;
+
+    if( semiblocks < MIN_SEMIBLOCKS_COUNT )
+    {
+        ret = MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+        goto cleanup;
+    }
+
+    memcpy( output, input, semiblocks * KW_SEMIBLOCK_LENGTH );
```

Comment:
You're copying `semiblocks` semiblocks to `output`, but `mbedtls_nist_kw_unwrap` calls this function with `semiblocks = in_len / 8` whereas the size of the output buffer is one semiblock less than the input size. This means that there's a 1-semiblock (8-byte) buffer overflow.

`mbedtls_nist_kw_unwrap` ends up shifting `output` one semiblock to the left, and this function treats the first semiblock as a special case. You should put the first semiblock, i.e. `A`, in a separate 8-byte buffer which is allocated in `mbedtls_nist_kw_unwrap` and passed to `unwrap` as an argument. Reduce offsets in `output` by 1. Change
```
        R -= KW_SEMIBLOCK_LENGTH;
        if( R < output + KW_SEMIBLOCK_LENGTH )
            R = output + ( semiblocks -1 ) * KW_SEMIBLOCK_LENGTH;
```
to
```
        if( R == output )
            R = A;
        else
            R -= KW_SEMIBLOCK_LENGTH;
```
Remove the shifting. Adjust all sizes and offsets as necessary.